synapse/storage/databases/main/events.py

```diff
@@ -364,6 +365,20 @@ def _persist_events_txn(
         min_stream_order = events_and_contexts[0][0].internal_metadata.stream_ordering
         max_stream_order = events_and_contexts[-1][0].internal_metadata.stream_ordering
 
+        # We check that the room still exists for events we're trying to
+        # persist. This is to protect against races with deleting a room.
+        #
+        # Annoyingly SQLite doesn't support row level locking.
+        if isinstance(self.database_engine, PostgresEngine):
+            for room_id in set(e.room_id for e, _ in events_and_contexts):
+                txn.execute(
+                    "SELECT room_version FROM rooms WHERE room_id = ? FOR SHARE",
+                    (room_id,),
+                )
```

Comment:
If `_purge_room_txn` locks the row first, I think we'll block until it's done, then we'll observe "ERROR:  could not serialize access due to concurrent update", retry, fail to find the room and raise.

I suppose it's only a problem if `_purge_room_txn` takes a long time to complete?

Reply:
Yup, basically. But I don't think its a *huge* problem to hold onto a transaction for that long, so long as we don't have many transactions like that.